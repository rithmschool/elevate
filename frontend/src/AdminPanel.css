.admin-panel{
  display: grid;
  grid-template-areas: "content side";
  grid-template-columns: 70% 30%;
  height: 100vh;
  grid-area: content;
  border: 1px solid black;
}

.admin-navbar {
  grid-area: side;
  border: 1px solid blue;
}

@media (max-width: 875px) {

}

@media only screen and (min-width:480px) and (max-width: 640px)  {
  .admin-panel{
    grid-template-columns: 100%;
  }

  .admin-navbar {
    display: none;
  }
}