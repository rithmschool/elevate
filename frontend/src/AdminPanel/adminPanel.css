.adminPanel_h1 {
  font-size: 2em;
}

.adminPanel_main {
  display: grid;
  grid-template-areas: "content side";
  grid-template-columns: 75% 25%;
  height: 100vh;
  background: #ECF0F5;
}

.adminPanel_panel {
  grid-area: content;
  font-size: 0.75rem;
  text-transform: capitalize;
  padding: 1.2em;
}

.adminPanel_navbar {
  grid-area: side;
  background: #222D32;
  color: lightgrey;
  padding: 0.7em;
}

@media only screen and (max-width: 640px)  {
  .adminPanel_main{
    grid-template-columns: 100%;
  }

  .adminPanel_navbar {
    display: none;
  }
}
