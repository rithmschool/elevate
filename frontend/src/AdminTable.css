.admin-table {
  background: white;
  font-size: 1em;
}

@media only screen and (max-width: 640px)  {
  .admin-table > thead th {
    max-width: 0.5em;
    background: green;
  }
}
