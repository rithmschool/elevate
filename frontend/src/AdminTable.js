import React, { Component } from 'react';
import { Table } from 'react-bootstrap';

class AdminTable extends Component {
  handleClick = (evt) => {
    this.props.getUserDetail(evt);
  }

  render(){
    const keys = Object.keys(this.props.tableObjs[0]);

    return (
      <Table striped bordered hover size="sm" responsive id="users-table">
        <thead>
          <tr>
            { keys.map(key => {
              key = key.replace(/_/g, ' ');
              return <th key={ key } width='55vw'>{ key }</th> 
            })}
          </tr>
        </thead>
        <tbody>
          {this.props.tableObjs.map(item => {
            const itemKeys = Object.keys(item)
            const itemValues = Object.values(item);
            return (
              <tr key={itemValues[0]} onClick={this.handleClick}>
                {itemValues.map((value, index) => {
                  return (
                    <td key={`${itemValues[0]}-${itemKeys[index]}`}>{ value }</td>
                  );
                })}
              </tr>
            );
          })}
        </tbody>
      </Table>
    )
  }
}

export default AdminTable;