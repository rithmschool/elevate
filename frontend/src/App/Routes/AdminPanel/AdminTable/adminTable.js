import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import { Table } from "react-bootstrap";

import "./adminTable.css";

const mql = window.matchMedia(`(max-width: 640px)`);
// set the maximum number of table colums for smaller and larger screens
const maxColumCount = mql.matches ? 5 : 12;

class AdminTable extends Component {
  handleClick = evt => {
    const id = evt.target.parentElement.id;
    this.props.history.push(`/admin/users/${id}`);
  };

  createTableHeader() {
    const keys = Object.keys(this.props.tableObjs[0]);

    return (
      <tr>
        {keys
          .map(key => {
            // Remove underscore from key name
            key = key.replace(/_/g, " ");

            return <th key={key}>{key}</th>;
          })
          .filter((key, idx) => idx < maxColumCount)}
      </tr>
    );
  }

  createTableBody() {
    const table = this.props.tableObjs.map(item => {
      const itemKeys = Object.keys(item);
      const itemValues = Object.values(item);
      return this.createTableRows(itemKeys, itemValues);
    });

    return table;
  }

  concantinateText(value) {
    // concatinate at 6 characters for small screen and 25 for large
    if (typeof value === "string") {
      if (mql.matches) {
        // small screen
        if (value.slice(0, 2) === "19" || value.slice(0, 2) === "20") {
          // only show year on small strings
          value = value.slice(0, 4);
        }
        if (value.length > 9) {
          value = value.slice(0, 6) + "...";
        }
      } else {
        // larger screen
        if (value.length > 30) {
          value = value.slice(0, 25) + "...";
        }
      }
    }

    return value;
  }

  createTableRows(keys, values) {
    return (
      <tr key={values[0]} onClick={this.handleClick} id={values[0]}>
        {values
          .map((value, index) => {
            value = this.concantinateText(value);

            return <td key={`${values[0]}-${keys[index]}`}>{value}</td>;
          })
          .filter((value, idx) => idx < maxColumCount)}
      </tr>
    );
  }

  render() {
    const tableType = this.props.tableObjs[0].hasOwnProperty("company")
      ? "users"
      : "questions";
    return (
      <div className="admin-table">
        <Table
          striped
          bordered
          hover
          size="sm"
          responsive
          id={tableType + "-table"}
        >
          <thead>{this.createTableHeader()}</thead>
          <tbody>{this.createTableBody()}</tbody>
        </Table>
      </div>
    );
  }
}

export default withRouter(AdminTable);
