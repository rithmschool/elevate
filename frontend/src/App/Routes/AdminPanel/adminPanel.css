.adminPanel_h1 {
  font-size: 2em;
}

.adminPanel_main {
  display: grid;
  grid-template-areas: "content side";
  height: 100vh;
  background: #ECF0F5;
}

.adminPanel_main_showing {
  grid-template-columns: 75% 25%;
}

.adminPanel_main_docked {
  grid-template-columns: 93% 7%;
}

.adminPanel_navbar {
  position: relative; 
  overflow: hidden;
  background: #222D32;
  color: lightgrey;
}

.adminPanel_navbar_showing {
}

.adminPanel_navbar_docked {
  cursor: pointer;
  padding: 0;
}

.adminPanel_panel {
  font-size: 0.75rem;
  text-transform: capitalize;
  padding: 1.2em;
}

@media only screen and (max-width: 640px)  {
  .adminPanel_main_showing {
    grid-template-columns: 30% 70%;
  }
  .adminPanel_main_docked {
    grid-template-columns: 83% 17%;
  }
}

@media only screen and (min-width: 1200px)  {
  .adminPanel_main_showing {
    grid-template-columns: 80% 20%;
  }
  .adminPanel_main_docked {
    grid-template-columns: 96% 4%;
  }
}
