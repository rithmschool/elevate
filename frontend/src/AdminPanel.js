import React, { Component } from 'react';
import './AdminPanel.css'
import AdminNavBar from './AdminNavBar';
import AdminUserView from './AdminUserView';
import AdminTable from './AdminTable';
import ElevateApi from './ElevateApi';

const mql = window.matchMedia(`(max-width: 640px)`);

class AdminPanel extends Component {
  constructor(props){
    super(props);

    this.state = {
      view: '',
      sidebarDocked: mql.matches,
      sideBarOpen: false,
      users: null,
      questions:null,
      userDetail: null
    }
  }

  componentDidMount = async () => {
    mql.addListener(this.mediaQueryChanged);
    let users;
    let questions;

    try {
      users = await ElevateApi.getUsers();
      questions = await ElevateApi.getQuestions();
    } catch(err) {
      console.log(err)
      return err;
    }
    
    this.setState({ users, questions });
  }

  // get update users after delete a user in AdminUserView
  updateUserState = (users) => {
    this.setState({users})
  }

  changeView = (view) => {
    this.setState({ view })
  }

  mediaQueryChanged = () => {
    this.setState({ sidebarDocked: mql.matches });
  }

  toggleSidebar = () => {
    this.setState({ sideBarOpen: !this.state.sideBarOpen });
  }

  getUserDetail = async (userId) => {
    const user = await ElevateApi.getUser(userId);
    
    this.setState({ view: 'userDetail', userDetail: user });
  }

  render(){
    if (!this.state.users || !this.state.questions){
      return (<div>...Loading</div>);
    }

    return(
      <div className="admin-main">
        <div className="admin-panel">
          { mql.matches && <button onClick={this.toggleSidebar}>SIDEBAR</button> }
          <h1 className="admin-h1">Admin Panel</h1>
          { this.state.sideBarOpen && <AdminNavBar changeView={this.changeView} /> }
          { this.state.view === 'users' || this.state.view === 'questions' ? 
            <AdminTable tableObjs={ this.state[this.state.view] } 
                        getUserDetail={ this.getUserDetail }
                        view={ this.state.view } /> : null
          }
          {this.state.view === 'userDetail' && 
                                <AdminUserView 
                                  user={this.state.userDetail} 
                                  updateUserState={this.updateUserState} 
                                  changeView={this.changeView} /> 
          }
        </div>
        
        <div className="admin-navbar">
          <AdminNavBar changeView={this.changeView}/>
        </div>
      </div>
    )
  }
}

export default AdminPanel;
