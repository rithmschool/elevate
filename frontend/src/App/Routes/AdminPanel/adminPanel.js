import React from "react";
import "./adminPanel.css";

import AdminNavbar from "./AdminNavbar/adminNavbar";
import AdminUserView from "./AdminUserView/adminUserView";
import AdminTable from "./AdminTable/adminTable";
import ElevateApi from "../../../elevateApi";
import Spinner from "../../../Spinner/spinner";
import PanelToggleBtn from "./PanelToggleBtn/panelToggleBtn";


class AdminPanel extends React.Component {
  constructor(props){
    super(props);

    this.state = {
      view: "",
      sideBarOpen: true,
      users: null,
      questions:null,
      userDetail: null
    }
  }

  componentDidMount = async () => {
    let users;
    let questions;

    try {
      users = await ElevateApi.getUsers();
      questions = await ElevateApi.getQuestions();
    } catch(err) {
      console.log(err)
      return err;
    }
    
    this.setState({ users, questions });
  }

  // get update users after delete a user in AdminUserView
  updateUserState = users => {
    this.setState({ users })
  }

  changeView = view => {
    this.setState({ view })
  }

  toggleSidebar = () => {
    this.setState(st => ({ sideBarOpen: !st.sideBarOpen }));
  }

  getUserDetail = async userId => {
    const user = await ElevateApi.getUser(userId);
    
    this.setState({ view: "userDetail", userDetail: user });
  }

  render(){
    if (!this.state.users || !this.state.questions){
      return <Spinner />;
    }

    const position = this.state.sideBarOpen ? "showing" : "docked";

    return (
      <div className={ `adminPanel_main adminPanel_main_${ position }` }>

        <div className="adminPanel_panel">

          <h1 className="adminPanel_h1">Admin Panel</h1>

          { ["users", "questions"].includes(this.state.view)
              && <AdminTable 
                   tableObjs={ this.state[this.state.view] } 
                   getUserDetail={ this.getUserDetail }
                   view={ this.state.view } /> }

          { this.state.view === "userDetail" 
              && <AdminUserView 
                   user={ this.state.userDetail }
                   updateUserState={ this.updateUserState }
                   changeView={ this.changeView } /> }
        </div>
        
        <div className={`adminPanel_navbar adminPanel_navbar_${ position }` }>
          <div>
            <PanelToggleBtn 
              toggleSidebar={this.toggleSidebar}
              direction={ this.state.sideBarOpen 
                ? 'toggleRight' 
                : 'toggleLeft' } />
          </div>

          <AdminNavbar 
            position={ this.state.sideBarOpen }
            toggleSidebar={ this.toggleSidebar }
            changeView={ this.changeView } />
        </div>

      </div>
    )
  }
}

export default AdminPanel;
