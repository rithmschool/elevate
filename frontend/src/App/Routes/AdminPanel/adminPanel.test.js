import React from "react";
import { MemoryRouter } from "react-router-dom";
import { shallow, mount } from "enzyme";
import mockData from "../../../../../mock_data";
import AdminPanel from "./adminPanel";
import AdminTable from "./AdminTable/adminTable";
import ElevateApi from "../../../elevateApi";

const { USERS, QUESTIONS } = mockData;
const users = USERS;
const questions = QUESTIONS;

describe("AdminPanel", function() {
  let wrapper;

  beforeEach(async () => {
    ElevateApi.getUsers = jest.fn(() => users);
    ElevateApi.getQuestions = jest.fn(() => questions);
    wrapper = mount(
      <MemoryRouter initialEntries={["/admin/users"]}>
        <AdminPanel
          history={{ location: { pathname: "/admin/users" } }}
          match={{ params: {} }}
        />
      </MemoryRouter>
    );
    wrapper.find(AdminPanel).setState({ users, questions });
  });

  it("renders without crashing", function() {
    shallow(<AdminPanel />);
  });

  it("makes API calls upon mounting", function() {
    expect(ElevateApi.getUsers).toHaveBeenCalled();
    expect(ElevateApi.getQuestions).toHaveBeenCalled();
  });

  it("toggles admin navbar on clicking toggle button", function() {
    const toggleBtn = wrapper.find("#adminPanel_toggleBtn");
    toggleBtn.simulate("click");
    expect(wrapper.find(AdminPanel).state("sideBarOpen")).toBe(false);
    toggleBtn.simulate("click");
    expect(wrapper.find(AdminPanel).state("sideBarOpen")).toBe(true);
  });

  it("has states", function() {
    let panel = wrapper.find(AdminPanel);
    expect(panel.state("sideBarOpen")).toEqual(true);
    expect(panel.state("users")).toEqual(users);
    expect(panel.state("questions")).toEqual(questions);
  });

  it("renders AdminPanel and AdminTable components", function() {
    wrapper.find(AdminPanel).setState({ users, questions });
    expect(wrapper.find(AdminPanel)).toHaveLength(1);
  });

  it("has div with adminPanel_main class", function() {
    expect(wrapper.find("div.adminPanel_main")).toHaveLength(1);
  });

  it("has div with adminPanel_panel class", function() {
    expect(wrapper.find("div.adminPanel_main")).toHaveLength(1);
  });
});
