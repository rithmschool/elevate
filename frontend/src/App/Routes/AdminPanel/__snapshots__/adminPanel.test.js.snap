// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPanel matches snapshot 1`] = `
<AdminPanel>
  <div
    className="adminPanel_main adminPanel_main_showing"
  >
    <div
      className="adminPanel_panel"
    >
      <h1
        className="adminPanel_h1"
      >
        Admin Panel
      </h1>
    </div>
    <div
      className="adminPanel_navbar adminPanel_navbar_showing"
    >
      <div>
        <PanelToggleBtn
          direction="toggleRight"
          toggleSidebar={[Function]}
        >
          <div
            className="panelToggleBtn"
          >
            <span
              id="adminPanel_toggleBtn"
              onClick={[Function]}
            >
              &gt;
            </span>
          </div>
        </PanelToggleBtn>
      </div>
      <AdminNavbar
        changeView={[Function]}
        position={true}
        toggleSidebar={[Function]}
      >
        <div
          className="adminNavbar"
        >
          <p
            id="users"
            onClick={[Function]}
          >
            Users
          </p>
          <p
            id="questions"
            onClick={[Function]}
          >
            Questions
          </p>
          <p
            id="invoices"
            onClick={[Function]}
          >
            Invoices
          </p>
          <p
            id="calendar"
            onClick={[Function]}
          >
            Calendar
          </p>
        </div>
      </AdminNavbar>
    </div>
  </div>
</AdminPanel>
`;
